crates/holochain/src/core/state/metadata.rs

```diff
@@ -512,12 +594,111 @@ where
         let key = ChainItemKey::from(header);
         let key = MiscMetaKey::chain_item(&key).into();
         let value = MiscMetaValue::ChainItem(header.timestamp().clone().into());
-        self.misc_meta.put(key, value)
+        self.misc_meta.put(key, value)?;
+        self.update_activity_status(header)
     }
 
     fn deregister_activity(&mut self, header: &Header) -> DatabaseResult<()> {
         let key = ChainItemKey::from(header);
-        self.misc_meta.delete(MiscMetaKey::chain_item(&key).into())
+        self.misc_meta
+            .delete(MiscMetaKey::chain_item(&key).into())?;
+        self.update_activity_status(header)
+    }
+
+    fn register_activity_status(
+        &mut self,
+        agent: &AgentPubKey,
+        status: ChainStatus,
+    ) -> DatabaseResult<()> {
+        // Rules to make this monotonic
+        // - Invalid overwrites valid and any invalids later in the chain.
+        // - Later Valid headers overwrite earlier Valid.
+        // - If there are two Valid status at the same seq num then insert an Invalid.
```

Comment:
It would be awesome to write some empty, ignored tests that "test" this behavior. Then we can at least get some "specs" written which can get fleshed out as actual tests later. Could start using the "TEST:" convention too. Like:

```rust
/// TEST: Invalid status overwrites valid and any invalid status later in the chain
#[test]
#[ignore = "test stub"]
fn test_register_activity_status_invalid_overwrites() {
    todo!("write test")
}

/// TEST: Later Valid headers overwrite earlier Valid
#[test]
#[ignore = "test stub"]
fn test_register_activity_status_valid_overwrites() {
    todo!("write test")
}


/// TEST: If there are two Valid status at the same seq num then insert an Invalid.
#[test]
#[ignore = "test stub"]
fn test_register_activity_status_forks() {
    todo!("write test")
}
```

Reply:
I'm a little hesitant to do this in this PR just because the size is already getting to large but I think it's a good idea

Reply:
I've added a PR to test this https://github.com/holochain/holochain/pull/431